src/commonMain/kotlin/com/dxc/ssi/agent/api/impl/SsiAgentApiImpl.kt

```
@@ -102,16 +126,22 @@ class SsiAgentApiImpl(
 
                 println("init: initialized trust ping tracker service")
 
-                libraryStateListener.initializationCompleted()
+                agentScope.launch { libraryStateListener.initializationCompleted() }
+
             } catch (t: Throwable) {
-                libraryStateListener.initializationFailed()
+                agentScope.launch {
+                    libraryStateListener.initializationFailed(
+                        LibraryError.LISTENER_SETUP_EXCEPTION,
+                        t.stackTraceToString()
+                    )
+                }
             }
 
         }
 
     }
 
-    override fun connect(url: String, keepConnectionAlive: Boolean): PeerConnection {
+    override fun connect(url: String, keepConnectionAlive: Boolean): PeerConnection? {
```

I think it's risky to change to PeerConnection? , the better pattern is to have not nullable and add a field inside to inform that is no longer needed / or use empty constructors. It can leads that other modules need to handle this information.

---

I agree that is not very good, but in some cases we can not return a connection
1) If we failed to parse invitation URL - it is avoidable, by moving parsing outside of this function
2) If we failed to store connection record in a wallet - solution for this would only be to throw an exception.

But now we are supposed to catch all exceptions inside of the library to avoid crashes on ios application. We could overcome this by using @Throws but I would prefer to do it in scope of separate feature, because we will have to make rework in all functions and massive retest

---

So I suggest to have nullable connection for now.

---

ok, thanks for the explanation.